faunadb-java-dsl/src/main/java/com/faunadb/client/types/Codecs.java

```diff
@@ -0,0 +1,103 @@
+package com.faunadb.client.types;
+
+import com.faunadb.client.types.Properties.Property;
+import com.faunadb.client.types.Value.ObjectV;
+import com.faunadb.client.types.time.HighPrecisionTime;
+import com.google.common.base.Function;
+import com.google.common.base.Functions;
+import org.joda.time.Instant;
+import org.joda.time.LocalDate;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.faunadb.client.types.Constructors.createDecoder;
+import static com.google.common.collect.Maps.newHashMap;
+import static java.lang.String.format;
+
+final class Codecs {
+  private Codecs() {}
+
+  private static final ConcurrentHashMap<Class<?>, Codec<?>> CODECS = new ConcurrentHashMap<>();
+
+  static {
+    CODECS.put(String.class, Codec.STRING);
+    CODECS.put(Boolean.class, Codec.BOOLEAN);
+    CODECS.put(boolean.class, Codec.BOOLEAN);
+    CODECS.put(Long.class, Codec.LONG);
+    CODECS.put(long.class, Codec.LONG);
+    CODECS.put(Integer.class, Codec.INTEGER);
+    CODECS.put(int.class, Codec.INTEGER);
+    CODECS.put(Short.class, Codec.SHORT);
+    CODECS.put(short.class, Codec.SHORT);
+    CODECS.put(Byte.class, Codec.BYTE);
+    CODECS.put(byte.class, Codec.BYTE);
+    CODECS.put(Character.class, Codec.CHAR);
+    CODECS.put(char.class, Codec.CHAR);
+    CODECS.put(Float.class, Codec.FLOAT);
+    CODECS.put(float.class, Codec.FLOAT);
+    CODECS.put(Double.class, Codec.DOUBLE);
+    CODECS.put(double.class, Codec.DOUBLE);
+    CODECS.put(LocalDate.class, Codec.DATE);
+    CODECS.put(HighPrecisionTime.class, Codec.HP_TIME);
+    CODECS.put(Instant.class, Codec.TIME);
+    CODECS.put(byte[].class, Codec.BYTES);
+  }
+
+  public static Codec<?> getCodec(Class<?> type) {
+    Codec<?> codec = CODECS.get(type);
+
+    if (codec == null) {
+      codec = new ObjectCodec(createDecoder(type), createEncoder(type));
+
+      CODECS.put(type, codec);
+    }
+
+    return codec;
+  }
+
+  private static Function<Object, Result<Value>> createEncoder(final Class<?> type) {
+    try {
+      final Property[] readProperties = Properties.getReadProperties(type);
+
+      return new Function<Object, Result<Value>>() {
+        @Override
+        public Result<Value> apply(Object value) {
+          Map<String, Value> fields = newHashMap();
+
+          for (Property property : readProperties) {
+            try {
+              fields.put(property.getName(), Encoder.encode(property.get(value)).get());
+            } catch (Exception ex) {
+              return Result.fail(format("Could not encode field `%s`. Reason: %s", property.getName(), ex.getMessage()));
+            }
+          }
+
+          return Result.<Value>success(new ObjectV(fields));
+        }
+      };
+    } catch (Exception ex) {
+      return (Function)Functions.constant(Result.fail(ex.getMessage()));
```

Comment:
What are the issues capture by this catch? If you can't create an encoder for the desired type, it represents a major issue with the developer's code and it should break the stack, IMO.